### ISSUE TITLE
Unseal `DiagnosticDescriptor`

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#62899]
## Background and Motivation

Currently, the only way to add custom information to a diagnostic descriptor is via custom tags, which is an `IEnumerable<string>`

## Proposed API

```diff
namespace Microsoft.CodeAnalysis
{
-     public sealed class DiagnosticDescriptor : IEquatable<DiagnosticDescriptor?>
+     public class DiagnosticDescriptor : IEquatable<DiagnosticDescriptor?>
```

## Usage Examples

The IDE *could* for example use an `internal class IdeDiagnosticDescriptor : DiagnosticDescriptor` that has extra information such as fading option, enforce on build values, etc.

Currently enforce on build values are encoded as custom tags. But fading options are set to the whole analyzer.
Regardless of whether the IDE team is interested to do things that way, I can see other 3rd party analyzers making use of custom descriptors.

## Risks

@mavasani @333fred What do you think?

**Comment by Youssef1313** [comment]
It wasn't `sealed` originally, and was sealed in https://github.com/dotnet/roslyn/pull/1065.

**Comment by mavasani** [comment]
@Youssef1313 I don't think this is possible. Descriptors are used to create Diagnostic instances, and `Diagnostic.Create` supports overloads which take individual fields of the descriptor, and we need to synthesize a Descriptor on the Diagnostic instance from these parameters, which won't be possible if entire content of the Descriptor is not known to the compiler layer (which will happen with your proposal).

https://github.com/dotnet/roslyn/blob/87bf0d0026d379f9434a2d6ee758aa25ed327ae2/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L295-L298

https://github.com/dotnet/roslyn/blob/87bf0d0026d379f9434a2d6ee758aa25ed327ae2/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L165-L278

This also ensures that the DiagnosticDescriptor can be round-tripped across processes by all Roslyn layers, including the compiler layer.

**Comment by Youssef1313** [comment]
> Diagnostic.Create supports overloads which take individual fields of the descriptor, and we need to synthesize a Descriptor on the Diagnostic instance from these parameters, which won't be possible if entire content of the Descriptor is not known to the compiler layer (which will happen with your proposal).

When using custom descriptors, we'll use the `Diagnostic.Create` overload that takes a `DiagnosticDescriptor`, not those that take the individual fields. Does this cause a problem?
